File: bindings/jvm/src/main/kotlin/nuventra/nuvexadb/NuvexaDatabase.kt
package nuventra.nuvexadb

import com.sun.jna.Pointer
import com.sun.jna.ptr.IntByReference
import com.sun.jna.ptr.LongByReference
import com.sun.jna.ptr.PointerByReference
import org.json.JSONArray
import org.json.JSONObject

class NuvexaDatabase private constructor(
    private val lib: NuvexaLibrary,
    private var handle: Long
) : AutoCloseable {

    fun insert(collection: String, json: String): String {
        val out = PointerByReference()
        check(lib.nuvexa_insert(handle, collection, json, out), out)
        return readAndFree(out)
    }

    fun insert(collection: String, document: NuvexaDocument): String =
        insert(collection, document.json.toString())

    fun replace(collection: String, json: String) {
        check(lib.nuvexa_replace(handle, collection, json))
    }

    fun deleteById(collection: String, id: String): Boolean {
        val deleted = IntByReference()
        check(lib.nuvexa_delete_by_id(handle, collection, id, deleted))
        return deleted.value != 0
    }

    fun findById(collection: String, id: String): NuvexaDocument? {
        val out = PointerByReference()
        val status = lib.nuvexa_find_by_id(handle, collection, id, out)
        if (status == NuvexaLibrary.NOT_FOUND) {
            free(out)
            return null
        }
        check(status, out)
        return NuvexaDocument.parse(readAndFree(out))
    }

    fun execute(nql: String): List<NuvexaDocument> {
        val out = PointerByReference()
        check(lib.nuvexa_execute(handle, nql, out), out)
        val array = JSONArray(readAndFree(out))
        return (0 until array.length()).map { NuvexaDocument(array.getJSONObject(it)) }
    }

    fun ensureIndex(collection: String, fields: List<String>) {
        val payload = if (fields.size == 1) JSONObject.quote(fields[0]) else JSONArray(fields).toString()
        check(lib.nuvexa_ensure_index(handle, collection, payload))
    }

    fun ensureIndex(collection: String, field: String) = ensureIndex(collection, listOf(field))

    fun insertMany(collection: String, jsonArray: String): List<String> {
        val out = PointerByReference()
        check(lib.nuvexa_insert_many(handle, collection, jsonArray, out), out)
        val array = JSONArray(readAndFree(out))
        return (0 until array.length()).map { array.getString(it) }
    }

    fun listCollections(): List<String> {
        val out = PointerByReference()
        check(lib.nuvexa_list_collections(handle, out), out)
        val array = JSONArray(readAndFree(out))
        return (0 until array.length()).map { array.getString(it) }
    }

    fun dropCollection(collection: String) {
        check(lib.nuvexa_drop_collection(handle, collection))
    }

    fun renameCollection(from: String, to: String) {
        check(lib.nuvexa_rename_collection(handle, from, to))
    }

    fun listIndexes(collection: String): List<JSONObject> {
        val out = PointerByReference()
        check(lib.nuvexa_list_indexes(handle, collection, out), out)
        val array = JSONArray(readAndFree(out))
        return (0 until array.length()).map { array.getJSONObject(it) }
    }

    fun dropIndex(collection: String, name: String) {
        check(lib.nuvexa_drop_index(handle, collection, name))
    }

    @JvmOverloads
    fun count(collection: String, filterJson: String = "{}"): Long {
        val n = LongByReference()
        check(lib.nuvexa_count(handle, collection, filterJson, n))
        return n.value
    }

    fun checkpoint() {
        check(lib.nuvexa_checkpoint(handle))
    }

    fun backup(destPath: String) {
        check(lib.nuvexa_backup(handle, destPath))
    }

    fun compact() {
        check(lib.nuvexa_compact(handle))
    }

    fun stats(): JSONObject {
        val out = PointerByReference()
        check(lib.nuvexa_stats(handle, out), out)
        return JSONObject(readAndFree(out))
    }

    fun changeEncryptionKey(currentKey: String, nextKey: String) {
        check(lib.nuvexa_change_encryption_key(handle, currentKey, nextKey))
    }

    fun beginTransaction() {
        check(lib.nuvexa_begin_transaction(handle))
    }

    fun commit() {
        check(lib.nuvexa_commit(handle))
    }

    fun rollback() {
        check(lib.nuvexa_rollback(handle))
    }

    @JvmOverloads
    fun uploadFile(fileName: String, sourcePath: String, chunkSize: Int = 0): String {
        val out = PointerByReference()
        check(lib.nuvexa_fs_upload(handle, fileName, sourcePath, chunkSize, out), out)
        return readAndFree(out)
    }

    fun downloadFile(fileId: String, destPath: String): Boolean {
        val found = IntByReference()
        check(lib.nuvexa_fs_download(handle, fileId, destPath, found))
        return found.value != 0
    }

    fun fileMetadata(fileId: String): NuvexaDocument? {
        val out = PointerByReference()
        val status = lib.nuvexa_fs_metadata(handle, fileId, out)
        if (status == NuvexaLibrary.NOT_FOUND) {
            free(out)
            return null
        }
        check(status, out)
        return NuvexaDocument.parse(readAndFree(out))
    }

    override fun close() {
        if (handle == 0L) {
            return
        }
        val status = lib.nuvexa_close(handle)
        handle = 0
        check(status)
    }

    private fun check(status: Int, out: PointerByReference? = null) {
        if (status == NuvexaLibrary.OK) {
            return
        }
        val message = lastError()
        free(out)
        throw when (status) {
            NuvexaLibrary.ENCRYPTION -> NuvexaEncryptionException(message)
            NuvexaLibrary.INTEGRITY -> NuvexaIntegrityException(message)
            else -> NuvexaException(message)
        }
    }

    private fun lastError(): String {
        val out = PointerByReference()
        lib.nuvexa_last_error(out)
        return readAndFree(out).ifBlank { "NuvexaDB native call failed." }
    }

    private fun readAndFree(out: PointerByReference): String {
        val pointer = out.value ?: return ""
        return try {
            pointer.getString(0, "UTF-8")
        } finally {
            lib.nuvexa_free(pointer)
        }
    }

    private fun free(out: PointerByReference?) {
        val pointer: Pointer = out?.value ?: return
        lib.nuvexa_free(pointer)
    }

    companion object {
        private val defaultLib: NuvexaLibrary by lazy { NuvexaLibrary.load() }

        @JvmStatic
        @JvmOverloads
        fun create(path: String, key: String? = null, library: NuvexaLibrary = defaultLib): NuvexaDatabase =
            openOrCreate(path, key, create = true, library)

        @JvmStatic
        @JvmOverloads
        fun open(path: String, key: String? = null, library: NuvexaLibrary = defaultLib): NuvexaDatabase =
            openOrCreate(path, key, create = false, library)

        @JvmStatic
        fun abiVersion(library: NuvexaLibrary = defaultLib): Int = library.nuvexa_abi_version()

        @JvmStatic
        @JvmOverloads
        fun restore(backupPath: String, destPath: String, overwrite: Boolean = false, library: NuvexaLibrary = defaultLib) {
            val status = library.nuvexa_restore(backupPath, destPath, if (overwrite) 1 else 0)
            if (status != NuvexaLibrary.OK) {
                val out = PointerByReference()
                library.nuvexa_last_error(out)
                val message = out.value?.getString(0, "UTF-8") ?: "NuvexaDB native call failed."
                if (out.value != null) {
                    library.nuvexa_free(out.value)
                }
                throw NuvexaException(message)
            }
        }

        @JvmStatic
        @JvmOverloads
        fun isEncrypted(path: String, library: NuvexaLibrary = defaultLib): Boolean {
            val flag = IntByReference()
            val status = library.nuvexa_is_encrypted(path, flag)
            if (status != NuvexaLibrary.OK) {
                val out = PointerByReference()
                library.nuvexa_last_error(out)
                val message = out.value?.getString(0, "UTF-8") ?: "NuvexaDB native call failed."
                if (out.value != null) {
                    library.nuvexa_free(out.value)
                }
                throw if (status == NuvexaLibrary.ENCRYPTION) {
                    NuvexaEncryptionException(message)
                } else {
                    NuvexaException(message)
                }
            }
            return flag.value != 0
        }

        private fun openOrCreate(
            path: String,
            key: String?,
            create: Boolean,
            library: NuvexaLibrary
        ): NuvexaDatabase {
            val handle = PointerByReference()
            val status = if (create) library.nuvexa_create(path, key, handle) else library.nuvexa_open(path, key, handle)
            if (status != NuvexaLibrary.OK) {
                val out = PointerByReference()
                library.nuvexa_last_error(out)
                val message = out.value?.getString(0, "UTF-8") ?: "NuvexaDB native call failed."
                if (out.value != null) {
                    library.nuvexa_free(out.value)
                }
                throw when (status) {
                    NuvexaLibrary.ENCRYPTION -> NuvexaEncryptionException(message)
                    NuvexaLibrary.INTEGRITY -> NuvexaIntegrityException(message)
                    else -> NuvexaException(message)
                }
            }
            return NuvexaDatabase(library, Pointer.nativeValue(handle.value))
        }
    }
}

File: bindings/swift/Sources/NuvexaDB/NuvexaDatabase.swift
import CNuvexa
import Foundation

public struct NuvexaDocument {
    public let json: String
    public let object: [String: Any]

    public var id: String { object["_id"] as? String ?? "" }

    public func field(_ name: String) -> String? {
        object[name].map { String(describing: $0) }
    }

    public static func parse(_ json: String) throws -> NuvexaDocument {
        guard let data = json.data(using: .utf8),
              let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NuvexaError.error("A NuvexaDB document must be a JSON object.")
        }
        return NuvexaDocument(json: json, object: object)
    }
}

public enum NuvexaError: Error, LocalizedError {
    case error(String)
    case encryption(String)
    case integrity(String)
    case notFound(String)

    public var errorDescription: String? {
        switch self {
        case .error(let m), .encryption(let m), .integrity(let m), .notFound(let m):
            return m
        }
    }

    static func from(status: Int32) -> NuvexaError {
        let message = nuvexaLastError()
        switch status {
        case Int32(NUVEXA_ENCRYPTION): return .encryption(message)
        case Int32(NUVEXA_INTEGRITY): return .integrity(message)
        case Int32(NUVEXA_NOT_FOUND): return .notFound(message)
        default: return .error(message)
        }
    }
}

public final class NuvexaDatabase {
    private var handle: nuvexa_handle

    public static func isEncrypted(_ path: String) throws -> Bool {
        var flag: Int32 = 0
        try path.withCString { cPath in
            try Self.check(nuvexa_is_encrypted(cPath, &flag))
        }
        return flag != 0
    }

    public static func create(_ path: String, key: String? = nil) throws -> NuvexaDatabase {
        try openOrCreate(path, key: key, create: true)
    }

    public static func open(_ path: String, key: String? = nil) throws -> NuvexaDatabase {
        try openOrCreate(path, key: key, create: false)
    }

    private static func openOrCreate(_ path: String, key: String?, create: Bool) throws -> NuvexaDatabase {
        var handle: nuvexa_handle = 0
        let status: Int32 = path.withCString { cPath in
            withOptionalCString(key) { cKey in
                create ? nuvexa_create(cPath, cKey, &handle) : nuvexa_open(cPath, cKey, &handle)
            }
        }
        try check(status)
        return NuvexaDatabase(handle: handle)
    }

    private init(handle: nuvexa_handle) {
        self.handle = handle
    }

    deinit {
        if handle != 0 {
            _ = nuvexa_close(handle)
        }
    }

    public func close() throws {
        guard handle != 0 else { return }
        let status = nuvexa_close(handle)
        handle = 0
        try Self.check(status)
    }

    @discardableResult
    public func insert(collection: String, json: String) throws -> String {
        var idPtr: UnsafeMutablePointer<CChar>?
        let status = collection.withCString { cCol in
            json.withCString { cJson in
                nuvexa_insert(handle, cCol, cJson, &idPtr)
            }
        }
        try Self.check(status)
        return nuvexaTakeString(&idPtr)
    }

    public func replace(collection: String, json: String) throws {
        let status = collection.withCString { cCol in
            json.withCString { cJson in
                nuvexa_replace(handle, cCol, cJson)
            }
        }
        try Self.check(status)
    }

    @discardableResult
    public func deleteById(collection: String, id: String) throws -> Bool {
        var deleted: Int32 = 0
        let status = collection.withCString { cCol in
            id.withCString { cId in
                nuvexa_delete_by_id(handle, cCol, cId, &deleted)
            }
        }
        try Self.check(status)
        return deleted != 0
    }

    public func findById(collection: String, id: String) throws -> NuvexaDocument? {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        let status = collection.withCString { cCol in
            id.withCString { cId in
                nuvexa_find_by_id(handle, cCol, cId, &jsonPtr)
            }
        }
        if status == Int32(NUVEXA_NOT_FOUND) {
            if let jsonPtr { nuvexa_free(jsonPtr) }
            return nil
        }
        try Self.check(status)
        return try NuvexaDocument.parse(nuvexaTakeString(&jsonPtr))
    }

    public func execute(_ nql: String) throws -> [NuvexaDocument] {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        let status = nql.withCString { nuvexa_execute(handle, $0, &jsonPtr) }
        try Self.check(status)
        let json = nuvexaTakeString(&jsonPtr)
        guard let data = json.data(using: .utf8),
              let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }
        return try array.map { object in
            let payload = try String(data: JSONSerialization.data(withJSONObject: object), encoding: .utf8)
                ?? "{}"
            return NuvexaDocument(json: payload, object: object)
        }
    }

    public func ensureIndex(collection: String, fields: [String]) throws {
        let payload = try String(data: JSONSerialization.data(withJSONObject: fields), encoding: .utf8) ?? "[]"
        let status = collection.withCString { cCol in
            payload.withCString { cFields in
                nuvexa_ensure_index(handle, cCol, cFields)
            }
        }
        try Self.check(status)
    }

    public func ensureIndex(collection: String, field: String) throws {
        try ensureIndex(collection: collection, fields: [field])
    }

    public static func abiVersion() -> Int32 {
        nuvexa_abi_version()
    }

    public static func restore(backupPath: String, destPath: String, overwrite: Bool = false) throws {
        let status = backupPath.withCString { cBackup in
            destPath.withCString { cDest in
                nuvexa_restore(cBackup, cDest, overwrite ? 1 : 0)
            }
        }
        try check(status)
    }

    public func insertMany(collection: String, jsonArray: String) throws -> [String] {
        var idsPtr: UnsafeMutablePointer<CChar>?
        let status = collection.withCString { cCol in
            jsonArray.withCString { cJson in
                nuvexa_insert_many(handle, cCol, cJson, &idsPtr)
            }
        }
        try Self.check(status)
        let json = nuvexaTakeString(&idsPtr)
        guard let data = json.data(using: .utf8),
              let array = try JSONSerialization.jsonObject(with: data) as? [String] else {
            return []
        }
        return array
    }

    public func listCollections() throws -> [String] {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        try Self.check(nuvexa_list_collections(handle, &jsonPtr))
        let json = nuvexaTakeString(&jsonPtr)
        guard let data = json.data(using: .utf8),
              let array = try JSONSerialization.jsonObject(with: data) as? [String] else {
            return []
        }
        return array
    }

    public func dropCollection(_ collection: String) throws {
        let status = collection.withCString { nuvexa_drop_collection(handle, $0) }
        try Self.check(status)
    }

    public func renameCollection(from: String, to: String) throws {
        let status = from.withCString { cFrom in
            to.withCString { cTo in
                nuvexa_rename_collection(handle, cFrom, cTo)
            }
        }
        try Self.check(status)
    }

    public func listIndexes(collection: String) throws -> [[String: Any]] {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        let status = collection.withCString { nuvexa_list_indexes(handle, $0, &jsonPtr) }
        try Self.check(status)
        let json = nuvexaTakeString(&jsonPtr)
        guard let data = json.data(using: .utf8),
              let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }
        return array
    }

    public func dropIndex(collection: String, name: String) throws {
        let status = collection.withCString { cCol in
            name.withCString { cName in
                nuvexa_drop_index(handle, cCol, cName)
            }
        }
        try Self.check(status)
    }

    public func count(collection: String, filterJson: String = "{}") throws -> Int64 {
        var n: Int64 = 0
        let status = collection.withCString { cCol in
            filterJson.withCString { cFilter in
                nuvexa_count(handle, cCol, cFilter, &n)
            }
        }
        try Self.check(status)
        return n
    }

    public func checkpoint() throws {
        try Self.check(nuvexa_checkpoint(handle))
    }

    public func backup(destPath: String) throws {
        let status = destPath.withCString { nuvexa_backup(handle, $0) }
        try Self.check(status)
    }

    public func compact() throws {
        try Self.check(nuvexa_compact(handle))
    }

    public func stats() throws -> [String: Any] {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        try Self.check(nuvexa_stats(handle, &jsonPtr))
        let json = nuvexaTakeString(&jsonPtr)
        guard let data = json.data(using: .utf8),
              let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [:]
        }
        return object
    }

    public func changeEncryptionKey(currentKey: String, nextKey: String) throws {
        let status = currentKey.withCString { cCurrent in
            nextKey.withCString { cNext in
                nuvexa_change_encryption_key(handle, cCurrent, cNext)
            }
        }
        try Self.check(status)
    }

    public func beginTransaction() throws {
        try Self.check(nuvexa_begin_transaction(handle))
    }

    public func commit() throws {
        try Self.check(nuvexa_commit(handle))
    }

    public func rollback() throws {
        try Self.check(nuvexa_rollback(handle))
    }

    public func uploadFile(fileName: String, sourcePath: String, chunkSize: Int32 = 0) throws -> String {
        var idPtr: UnsafeMutablePointer<CChar>?
        let status = fileName.withCString { cName in
            sourcePath.withCString { cPath in
                nuvexa_fs_upload(handle, cName, cPath, chunkSize, &idPtr)
            }
        }
        try Self.check(status)
        return nuvexaTakeString(&idPtr)
    }

    public func downloadFile(fileId: String, destPath: String) throws -> Bool {
        var found: Int32 = 0
        let status = fileId.withCString { cId in
            destPath.withCString { cDest in
                nuvexa_fs_download(handle, cId, cDest, &found)
            }
        }
        try Self.check(status)
        return found != 0
    }

    public func fileMetadata(fileId: String) throws -> NuvexaDocument? {
        var jsonPtr: UnsafeMutablePointer<CChar>?
        let status = fileId.withCString { nuvexa_fs_metadata(handle, $0, &jsonPtr) }
        if status == Int32(NUVEXA_NOT_FOUND) {
            if let jsonPtr { nuvexa_free(jsonPtr) }
            return nil
        }
        try Self.check(status)
        return try NuvexaDocument.parse(nuvexaTakeString(&jsonPtr))
    }

    private static func check(_ status: Int32) throws {
        if status != Int32(NUVEXA_OK) {
            throw NuvexaError.from(status: status)
        }
    }

}

private func nuvexaLastError() -> String {
    var ptr: UnsafeMutablePointer<CChar>?
    _ = nuvexa_last_error(&ptr)
    return nuvexaTakeString(&ptr)
}

private func nuvexaTakeString(_ pointer: inout UnsafeMutablePointer<CChar>?) -> String {
    guard let pointer else { return "" }
    defer { nuvexa_free(pointer) }
    return String(cString: pointer)
}

private func withOptionalCString<T>(_ value: String?, _ body: (UnsafePointer<CChar>?) -> T) -> T {
    guard let value else {
        return body(nil)
    }
    return value.withCString(body)
}
